core/src/main/scala/kafka/server/MetadataCache.scala

```diff
@@ -151,8 +158,17 @@ private[server] class MetadataCache(brokerId: Int) extends Logging {
     }
   }
 
+  def getControllerId: Int = {
+    controllerId.getOrElse(MetadataResponse.NO_CONTROLLER_ID)
+  }
+
   def updateCache(correlationId: Int, updateMetadataRequest: UpdateMetadataRequest) {
     inWriteLock(partitionMetadataLock) {
+      controllerId =
+        updateMetadataRequest.controllerId match {
+          case id if id < 0 => None
```

Comment:
The issue with this approach is that we accept instead of failing when we receive invalid values. In some cases, this can lead to bugs being missed.